# RpNGS: an interactive web platform for pathogen report generation using metagenomics data


# Summary
RpNGS is a graphical interface application for clinical metagenomics detection including basic data collection in the microbiology laboratory and sequencing data analysis in the computation lab. Experimenter, bioinformatician and clinician can work together for giving a pathogen detection report based on the clinical infomation and the interpretation of mNGS data, using interactive data tables, alignment coverage viewer and word template. Additionally, the analysis pipeline of application enables automated transfer fastq files from sequencer to server. 

## To run Shiny app:

Open Rstudio, and run below command 

> library("shiny“）
>
> runGitHub(repo = "RpNGS",username="mj200921059",ref = "main",subdir="RpNGS.R")

## The general workflow is described in below.

**1. Summary dataset**

---


![alt text](image.png)

The summary tab empowers users to visually analyze the accomplished mNGS test. The interactive visualization choice is a map, pie chart, and bar plot to illustrate the sales volume among locations, percentage of each sample type and samples size distribution between months in one year, respectively. RpNGS use main data table to display detail information of processed mNGS test for feasible searching and double checking the clinical reports. Click sample id in the first column will open the clinical metagenomics report word document or give a hint for report generation.

**2. Analyze dataset**

---

![image](https://github.com/user-attachments/assets/3ac9b75a-e898-4e27-a367-9945c8799445)


In RpNGS second tab, the trained experimenter should update the information of each batch including flow cell id, sample id, nucleic acids concentration after extraction and library preparation steps, adaptor id, and file name of sequencing data. Then start the process step by click the process button. There are six steps inside the process pipeline including detecting and copying the sequencing data from sequencer to server, quality control, host reads removal, and classification, abundance estimate, mapped reads extraction. To assist with distinguishing reads from microbes existing in the reagent and lab’s environment, RpNGS calculated z-scores of taxons in each sample by compared relative abundance estimations of specie to water-only or other control sample collections. The progress bar will track the status of five essential processes during data analyze.

**3. Test report**

---

![image](https://github.com/user-attachments/assets/7d2a5a8c-24df-46bb-93ec-9da39a111bd0)


The third tab shows clinician or certified user manually decide the pathogen for specific patient from microbes list based on multiple factors such as z-scores, mapped reads, average coverage for a specific pathogen, gender, age, sample types, clinical concerned pathogens types, anti-infection treatment. User should input the flow cell id for obtaining test samples in a status table with three columns, flow cell id, sample id and status columns. The sample’ clinical details and raw microbes list are activated based on the selection in the report status table. To assist with pathogen detection, the background color of microbes listed in all potential pathogen candidates retrieved from 132 clinical metagenomics reports, generated by Shaanxi Lifegen Co., Ltd, are filled with green. The alignment view and related data of specific species will emerge by choosing one of these colored pathogens in raw microbes list table. Finally, the pathogen table will be generated by multiple selections in the raw microbe’s table. A key feature includes the ability to export experiment data, patient basic information, confirmed pathogens and related features of corresponding pathogen in a Word document as clinical pathogen test report for each sample.




